x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/StringProcessor.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression.function.scalar.string;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.sql.SqlIllegalArgumentException;
+import org.elasticsearch.xpack.sql.expression.function.scalar.processor.runtime.Processor;
+import org.elasticsearch.xpack.sql.expression.function.scalar.whitelist.InternalSqlScriptUtils;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.function.Function;
+
+public class StringProcessor implements Processor {
+    
+    private interface StringFunction<R> {
+        default R apply(Object o) {
+            if (!(o instanceof String || o instanceof Character)) {
+                throw new SqlIllegalArgumentException("A string/char is required; received [{}]", o);
+            }
+
+            return doApply(o.toString());
+        }
+
+        R doApply(String s);
+    }
+
+    private interface NumericFunction<R> {
+        default R apply(Object o) {
+            if (!(o instanceof Number)) {
+                throw new SqlIllegalArgumentException("A number is required; received [{}]", o);
+            }
+
+            return doApply((Number) o);
+        }
+
+        R doApply(Number s);
+    }
+
+    public enum StringOperation {
+        ASCII((String s) -> s.length() == 0 ? null : Integer.valueOf(s.charAt(0))),
+        CHAR((Number n) -> {
+            int i = n.intValue();
+            return i < 0 || i > 255 ? null : String.valueOf((char) i);
+        }),
+        LCASE((String s) -> s.toLowerCase(Locale.ROOT)),
+        UCASE((String s) -> s.toUpperCase(Locale.ROOT)),
+        LENGTH((String s) -> StringFunctionUtils.trimTrailingWhitespaces(s).length()),
+        RTRIM((String s) -> StringFunctionUtils.trimTrailingWhitespaces(s)),
+        LTRIM((String s) -> StringFunctionUtils.trimLeadingWhitespaces(s)),
+        SPACE((Number n) -> {
+            int i = n.intValue();
+            if (i < 0) {
+                return null;
+            };
+            char[] spaces = new char[i];
+            char whitespace = ' ';
+            Arrays.fill(spaces, whitespace);
+            
+            return new String(spaces);
+        }),
+        BIT_LENGTH((String s) -> s.getBytes(StandardCharsets.UTF_8).length * 8),
```

Comment:
the `getBytes` should be replaced by `UnicodeUtils#calcUTF16toUTF8Length` instead.